### DISCUSSION TITLE
tsh login does not work behind AWS ALB

### DISCUSSION THREAD

**Discussion by yzislin** [gravitational/teleport#14228] [Q&A] [answered]
Hi,

I am hard time getting tsh login or tsh ssh to work with AWS deployment. It seems that this is relatively common scenario but it doesnt work. So I have a teleport master (ie proxy, web ui). It is running on AWS EC2 instance. I have created AWS ALB with a Amazon signed public certificate. Teleport master itself, is running on self signed cert. I have teleport nodes (ie other ec2 instances). All of them are using self signed certs for communication. I have configured --insecure flag for all of the teleport nodes/master to make this work. Everything works just fine. Users connect to the ALB, cert is valid, login works just fine, SSH to nodes through teleport works great, Applications via teleport UI work just fine. 
BUT there is a problem. I have users who want to be able to connect to EC2 instance (ie teleport nodes) via their terminal screen.
So I've read on your site that I should use TSH LOGIN and then TSH SSH. It does not work. I am using v2 in the config. From what I understand, TLS Routing is having hard time dealing with ALB and Teleport Master self signed cert. Here is the log from TSH LOGIN command:
`$ tsh login -d --insecure --proxy=teleport.mycompany.com:443 --user=teleport-admin
DEBU [CLIENT]    open /home/user/.tsh/teleport.mycompany.com.yaml: no such file or directory client/api.go:1009
INFO [CLIENT]    no host login given. defaulting to user client/api.go:1361
INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/tmp/ssh-XXXXXX92nGfy/agent.836" client/api.go:3404
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.mycompany.com:443 client/api.go:3365
DEBU             Attempting GET teleport.mycompany.com:443/webapi/ping webclient/webclient.go:118
Enter password for Teleport user teleport-admin:
Enter your OTP token:
xxxxxx
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.mycompany.com:443 client/api.go:3365
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.mycompany.com:443, insecure=true) client/weblogin.go:233
WARNING: You are using insecure connection to SSH proxy https://teleport.mycompany.com:443
DEBU [KEYAGENT]  Adding CA key for mycompany client/keyagent.go:324
DEBU [KEYSTORE]  Adding known host mycompany with proxy teleport.mycompany.com and key: SHA256:hzPndCxxxxxxBGVLXNlWB9twtfktC9CW5Orp4A client/keystore.go:578
DEBU [KEYSTORE]  Returning Teleport TLS certificate "/home/user/.tsh/keys/teleport.mycompany.com/teleport-admin-x509.pem" valid until "2022-06-22 03:20:13 +0000 UTC". client/keystore.go:307
DEBU [KEYAGENT]  Deleting obsolete stored key with index {ProxyHost:teleport.mycompany.com Username:teleport-admin ClusterName:mycompany}. client/keyagent.go:525
INFO [KEYAGENT]  Loading SSH key for user "teleport-admin" and cluster "mycompany". client/keyagent.go:202
INFO [CLIENT]    Connecting to proxy=teleport.mycompany.com:443 login="ubuntu" using TLS Routing client/api.go:2552
DEBU [KEYSTORE]  Returning Teleport TLS certificate "/home/kali/.tsh/keys/teleport.mycompany.com/teleport-admin-x509.pem" valid until "2022-07-09 02:26:23 +0000 UTC". client/keystore.go:307
DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:276

ERROR REPORT:
Original Error: x509.UnknownAuthorityError x509: certificate signed by unknown authority
Stack Trace:
        /go/src/github.com/gravitational/teleport/lib/utils/proxy/proxy.go:69 github.com/gravitational/teleport/lib/utils/proxy.directDial.dialALPNWithDeadline
        /go/src/github.com/gravitational/teleport/lib/utils/proxy/proxy.go:106 github.com/gravitational/teleport/lib/utils/proxy.directDial.Dial
        /go/src/github.com/gravitational/teleport/lib/client/api.go:2602 github.com/gravitational/teleport/lib/client.makeProxySSHClientWithTLSWrapper
        /go/src/github.com/gravitational/teleport/lib/client/api.go:2553 github.com/gravitational/teleport/lib/client.makeProxySSHClient
        /go/src/github.com/gravitational/teleport/lib/client/api.go:2525 github.com/gravitational/teleport/lib/client.(*TeleportClient).connectToProxy
        /go/src/github.com/gravitational/teleport/lib/client/api.go:2445 github.com/gravitational/teleport/lib/client.(*TeleportClient).ConnectToProxy.func1
        /opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
User Message: x509: certificate signed by unknown authority
`

So someone has advised me to set two settings in the proxy_service section of the config which I did.
Here is the full config for Teleport Master:
`version: v2
teleport:
  nodename: ip-xxx-xxx-xxx-xxx.ec2.internal
  data_dir: /opt/teleport
  log:
    output: /var/log/teleport/teleport.log
    severity: INFO
    format:
      output: text
  ca_pin: []
  diag_addr: ""
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: mycompany
  proxy_listener_mode: multiplex
  keep_alive_interval: 40s
  keep_alive_count_max: 3
ssh_service:
  enabled: "yes"
  labels:
    host: teleport
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:443
  public_addr: teleport.mycompany.com:443
  ssh_public_addr: public_ip_of_teleport_master:3023
  tunnel_public_addr: public_ip_of_teleport_master:3024
  https_keypairs:
  - key_file: /opt/teleport/server.key
    cert_file: /opt/teleport/server.cert
  acme: {}
`
The command I use to start the server is:
`/usr/local/bin/teleport start --pid-file=/run/teleport.pid --insecure -c /opt/teleport/teleport.yaml
`

It seems that this should work but it doesnt.

I feel like this is a very common scenario if teleport is used as a bastion host solution for AWS resources.

Please advise.

Thanks.

**Comment by webvictim** [comment, marked as answer]
~~You must choose either to use AWS ACM **or** TLS routing/multiplexing. You **CANNOT** use both. This is a technical limitation.~~

~~- For AWS ACM to work correctly, TLS routing/multiplexing must be disabled.~~

~~`tsh login` requires ports 443 and 3023 by default. 3023 should be a TCP passthrough direct to the Teleport proxy, probably using an NLB or ELB. 443 can be an HTTPS listener on an ALB with an ACM cert.~~

~~- For TLS routing/multiplexing to work, Teleport must terminate its own TLS - it must have both the certificate and private key.~~

~~Public AWS ACM does not provide you with the private key, so this isn't possible. You would need to get a certificate from an ACME provider (Let's Encrypt/ZeroSSL), a commercial certificate provider (Digicert etc) or use AWS Private CA with ACM. You could then provide the certificate to Teleport using `https_keypairs` and you wouldn't need ACM at all - you could use an NLB on port 443 which just passes TCP traffic through to Teleport.~~

Edit: This is no longer the case as of Teleport 13. Deploying Teleport in TLS multiplexing mode behind a TLS-terminating load balancer (such as an AWS LB with an ACM certificate) **is** now supported. You must be using `tsh` and `teleport` version 13+ for this to work. Teleport will automatically use ALPN connection upgrades to pass its traffic through the TLS termination.

More details on how this works can be found here: https://goteleport.com/docs/architecture/tls-routing/#working-with-layer-7-load-balancers-or-reverse-proxies

**Reply by yzislin** [reply]
Hold on a second. What about this scenario. 
Regular users still go to ALB with ACM cert and get their communication done through TLS (443). The users who want to TSH LOGIN/ TSH SSH, can communicate directly with Teleport Master over port 3023. teleport config on that server would have different addresses for proxy_service. There are a bunch of addr settings that can be configured for different scenarios. Can it work for my proposed solution?

This way, multiplexing is in place, ALB is used for regular users and other users can still do direct ssh.

Thanks.

**Reply by webvictim** [reply]
Yes, any web UI user that doesn’t need `tsh` can solely use port 443. This scenario works - but by its very definition it is not using multiplexing as there is more than one port involved.

**Reply by yzislin** [reply]
I understand. I just have couple of users who wants to use their linxu terminals and access servers over SSH instead of going through UI.

But I cant seem to get this TSH login to work still. Can you advise me on which settings I need to set in proxy_service and what --proxy command to use for TSH login?

Thank you.

**Comment by webvictim** [comment]
Here’s what your config file should look like to do this:

```yaml
teleport:
  nodename: ip-xxx-xxx-xxx-xxx.ec2.internal
  data_dir: /opt/teleport
  log:
    output: /var/log/teleport/teleport.log
    severity: INFO
    format:
      output: text
  ca_pin: []
  diag_addr: ""
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: mycompany
  keep_alive_interval: 40s
  keep_alive_count_max: 3
ssh_service:
  enabled: "yes"
  labels:
    host: teleport
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:443
  public_addr: teleport.mycompany.com:443
  ssh_public_addr: public_ip_of_teleport_master:3023
  tunnel_public_addr: public_ip_of_teleport_master:3024
  https_keypairs:
  - key_file: /opt/teleport/server.key
    cert_file: /opt/teleport/server.cert
  acme: {}
```

Make sure you have opened 443, 3023 and 3024 on the firewall for incoming traffic.

**Reply by yzislin** [reply]
So the problem is that I already have my config configured like that :)

It seems it tries to connect to teleport master IP and after password and OTP, it switches back to ALB for AWS. (teleport.mycompany.com is ALB endpoint).

Here is my command and the output:
`# tsh login -d --insecure  --proxy=public_ip_of_teleport_master:443 --user=teleport-admin
DEBU [CLIENT]    open /root/.tsh/public_ip_of_teleport_master.yaml: no such file or directory client/api.go:1016
INFO [CLIENT]    no host login given. defaulting to root client/api.go:1368
ERRO [CLIENT]    [KEY AGENT] Unable to connect to SSH agent on socket: "". client/api.go:3564
DEBU [CLIENT]    not using loopback pool for remote proxy addr: public_ip_of_teleport_master:443 client/api.go:3529
DEBU             Attempting GET public_ip_of_teleport_master:443/webapi/ping webclient/webclient.go:118
Enter password for Teleport user teleport-admin:
Enter your OTP token:
xxxxx
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.mycompany.com:443 client/api.go:3529
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.mycompany.com:443, insecure=true) client/weblogin.go:233
WARNING: You are using insecure connection to SSH proxy https://teleport.mycompany.com:443
DEBU [KEYAGENT]  Adding CA key for mycompany client/keyagent.go:324
DEBU [KEYSTORE]  Adding known host mycompany with proxy teleport.mycompany.com and key: SHA256:hzPndCQt4VJJmBIs0BGxxxxWB9twtfktC9CW5Orp4A client/keystore.go:578
DEBU [KEYSTORE]  Returning Teleport TLS certificate "/root/.tsh/keys/teleport.mycompany.com/teleport-admin-x509.pem" valid until "2022-06-22 06:37:54 +0000 UTC". client/keystore.go:307
DEBU [KEYAGENT]  Deleting obsolete stored key with index {ProxyHost:teleport.mycompany.com Username:teleport-admin ClusterName:mycompany}. client/keyagent.go:525
INFO [KEYAGENT]  Loading SSH key for user "teleport-admin" and cluster "mycompany". client/keyagent.go:202
INFO [CLIENT]    Connecting to proxy=teleport.mycompany.com:443 login="ubuntu" using TLS Routing client/api.go:2716
DEBU [KEYSTORE]  Returning Teleport TLS certificate "/root/.tsh/keys/teleport.mycompany.com/teleport-admin-x509.pem" valid until "2022-07-11 06:30:07 +0000 UTC". client/keystore.go:307
DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:276

ERROR REPORT:
Original Error: x509.UnknownAuthorityError x509: certificate signed by unknown authority
Stack Trace:
	/go/src/github.com/gravitational/teleport/lib/utils/proxy/proxy.go:69 github.com/gravitational/teleport/lib/utils/proxy.directDial.dialALPNWithDeadline
	/go/src/github.com/gravitational/teleport/lib/utils/proxy/proxy.go:106 github.com/gravitational/teleport/lib/utils/proxy.directDial.Dial
	/go/src/github.com/gravitational/teleport/lib/client/api.go:2766 github.com/gravitational/teleport/lib/client.makeProxySSHClientWithTLSWrapper
	/go/src/github.com/gravitational/teleport/lib/client/api.go:2717 github.com/gravitational/teleport/lib/client.makeProxySSHClient
	/go/src/github.com/gravitational/teleport/lib/client/api.go:2689 github.com/gravitational/teleport/lib/client.(*TeleportClient).connectToProxy
	/go/src/github.com/gravitational/teleport/lib/client/api.go:2609 github.com/gravitational/teleport/lib/client.(*TeleportClient).ConnectToProxy.func1
	/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
User Message: x509: certificate signed by unknown authority
`

**Reply by yzislin** [reply]
Sorry. I just realized that I've missed a line. You have removed proxy_listener_mode config line.
Does that mean that regular users who come through web ui, will experience difference in operation?

**Reply by yzislin** [reply]
Oh and do I have to know allow more ports between teleport proxy and EC2 instances? for SSH and for Applications?

Thanks.

**Reply by yzislin** [reply]
For any server when I try to SSH via Web UI (after authentication), I get this:
`disconnected
failed to authenticate with proxy localhost:3023
        dial tcp 127.0.0.1:3023: connect: connection refused`

**Reply by webvictim** [reply]
I also removed the `version: v2` line which is relevant.

You shouldn’t need any ports other than 443, 3023 and 3024 for SSH and applications.

**Reply by yzislin** [reply]
Thank you for your replies. I think I am going to go with letsencrypt option and remove ALB from the picture. I love multiplexing mode and just using one port.
